Don't give children negative sizes when overconstrained

eseidelGoogle opened the pull request:
We probably need a test here. Not sure how.  @Hixie 
The game was hitting this when in landscape mode.  @vlidholt

Hixie commented on packages/flutter/lib/src/rendering/flex.dart:
```diff
@@ -246,7 +247,7 @@ class RenderFlex extends RenderBox with ContainerRenderObjectMixin<RenderBox, Fl
         int flex = _getFlex(child);
         if (flex > 0) {
-          double childMainSize = spacePerFlex * flex;
+          double childMainSize = math.max(0.0, spacePerFlex * flex);
```
after looking at this more, I think your original intuition of having the math.max on the line that sets `spacePerFlex` is correct. It matches similar logic elsewhere.

Hixie commented:
To test thing create a file that has a sized box that contains a flex that contains a sized box bigger than the parent, and then contains a flexible containing a container. Then, expect() that the container's rendor object has a height of zero.

eseidelGoogle committed:
Fix two bugs in Flex exposed by the Game in landscape mode

One bug was masking the other, hence they both needed to be fixed
and tested separately.

@Hixie

eseidelGoogle commented:
ptal

Hixie commented on packages/flutter/test/rendering/flex_test.dart:
```diff
@@ -19,6 +19,65 @@ void main() {
+    );
+    RenderFlex flex = new RenderFlex(
+      direction: FlexDirection.vertical,
+      children: [
```
`<RenderBox>[...]`

Hixie commented on packages/flutter/test/rendering/flex_test.dart:
```diff
@@ -19,6 +19,65 @@ void main() {
+    );
+    RenderFlex flex = new RenderFlex(
+      direction: FlexDirection.horizontal,
+      children: [
```
`<RenderBox>[...]`

Hixie commented on packages/flutter/test/rendering/flex_test.dart:
```diff
@@ -19,6 +19,65 @@ void main() {
+  test('Vertical Flipped Constraints', () {
+    RenderFlex flex = new RenderFlex(
+      direction: FlexDirection.vertical,
+      children: [
```
`<RenderBox>[...]`

Hixie commented:
lgtm
